improver/regrid/bilinear.py

```diff
@@ -0,0 +1,507 @@
+    lat_spacing = in_latlons[in_lons_size, 0] - in_latlons[0, 0]
+    lon_spacing = in_latlons[1, 1] - in_latlons[0, 1]
+
+    # Calculate input/output offset, expressed in terms of the spacing
+    n_lat = (out_latlons[:, 0] - in_latlons[0, 0]) // lat_spacing
+    m_lon = (out_latlons[:, 1] - in_latlons[0, 1]) // lon_spacing
+    n_lat = n_lat.astype(int)
+    m_lon = m_lon.astype(int)
+
+    # Four surrounding input points for each output point, in a rectangle shape
+    index0 = n_lat * in_lons_size + m_lon
+    index1 = index0 + in_lons_size
+    index2 = index1 + 1
+    index3 = index0 + 1
+
+    # Rearrange order to match expected output style
+    # note: lat (X) but ordering  (lat0, lon0)(lat0,lon1)....(lat0,lon_last),(lat1,lon0),
+    indexes = np.transpose([index0, index1, index2, index3])
+    return indexes
+
+
+def basic_weights(index_range, indexes, out_latlons, in_latlons, in_lons_size):
+    """
+    calculate weighting for selecting target points using standard bilinear function
+    Args:
+        index_range((numpy.ndarray):
+            a list of target points
+        indexes(numpy.ndarray):
+            array of source grid point number for all target grid points
+        in_latlons(numpy.ndarray):
+            source points's latitude-longitudes
+        out_latlons(numpy.ndarray):
+            target points's latitude-longitudes
+        in_lons_size (int):
+            source grid's longitude dimension
+
+    Returns:
+        numpy.ndarray:
+            weigting array of source grid point number for target grid points
+    """
+    # Set up input points spacing values
+    lat_spacing = in_latlons[in_lons_size, 0] - in_latlons[0, 0]
+    lon_spacing = in_latlons[1, 1] - in_latlons[0, 1]
+    latlon_area = lat_spacing * lon_spacing
+
+    out_lats = out_latlons[index_range, 0]
+    out_lons = out_latlons[index_range, 1]
+    # Calculate weights for four surrounding input points
+    # Input point 1 - left bottom input point
+    lat_1 = in_latlons[indexes[index_range, 0], 0]
+    lon_1 = in_latlons[indexes[index_range, 0], 1]
+    # Input point 2 - right bottom input point
+    lat2_lat = lat_1 + lat_spacing - out_lats
+    # Input point 3 - right top input point
+    lon3_lon = lon_1 + lon_spacing - out_lons
+    lon_lon1 = out_lons - lon_1
+    lat_lat1 = out_lats - lat_1
+    # Input point 4 (left top) is implied by the others
+
+    weight1 = lat2_lat * lon3_lon / latlon_area
+    weight2 = lat_lat1 * lon3_lon / latlon_area
+    weight3 = lat_lat1 * lon_lon1 / latlon_area
+    weight4 = lat2_lat * lon_lon1 / latlon_area
+    weights = np.transpose([weight1, weight2, weight3, weight4])
+    return weights
+
+
+def adjust_for_surface_mismatch(
+    in_latlons,
+    out_latlons,
+    in_classified,
+    out_classified,
+    weights,
+    indexes,
+    surface_type_mask,
+    in_lons_size,
+    vicinity,
+):
+    """
+    updating source points and weighting for mismatched-source-point cases
```

Comment:
A lot more detail could be added to this docstring to make it easier to what calculations this function is trying to do. Even just bullet points outlining the main steps taken in the function to deal with different cases would help

Reply:
As we discussed, there are several pages of algorithm description. It is better to put somewhere else. Basically it is a science+software  project.  I reckon that we should not put science part into the code.

Reply:
Add 11-line summary of the algorithm

Reply:
Thank you